topology/probes/k8s/linker.go

```diff
@@ -0,0 +1,151 @@
+
+// AreLinked return true if (a, b) should be linked
+type AreLinked func(a, b interface{}) bool
+
+// ABLinker basis for a simple A to B linker
+type ABLinker struct {
+	Manager   string
+	Type      string
+	Graph     *graph.Graph
+	ACache    *ResourceCache
+	BCache    *ResourceCache
+	AreLinked AreLinked
+}
+
+// ABLinkerInterface returns the embbed ABLinker
+type ABLinkerInterface interface {
+	GetABLinker() *ABLinker
+}
+
+// GetABLinker returns inner object
+func (l *ABLinker) GetABLinker() *ABLinker {
+	return l
+}
+
+// NewEdgeMetadata create a new edge metadata
+func (l *ABLinker) NewEdgeMetadata() graph.Metadata {
+	return NewEdgeMetadata(l.Manager, l.Type)
+}
+
+// GenEdgeID generate a new ID for edge
+func (l *ABLinker) GenEdgeID(aNode, bNode *graph.Node) graph.Identifier {
+	return graph.GenID(string(aNode.ID), string(bNode.ID), "RelationType", l.Type)
+}
+
+// NewEdge create a new edge
+func NewEdgeIfMissing(g *graph.Graph, i graph.Identifier, parent, child *graph.Node, m graph.Metadata) *graph.Edge {
+	if g.GetEdge(i) != nil {
+		return nil
+	}
+	edge, err := g.NewEdge(i, parent, child, m, "")
+	if err != nil {
+		logging.GetLogger().Error(err)
+		return nil
+	}
+	return edge
+}
+
+// AppendNewEdge create and append a new edge to list of edges
+func (l *ABLinker) AppendNewEdge(inputEdges []*graph.Edge, aNode, bNode *graph.Node) []*graph.Edge {
+	i := l.GenEdgeID(aNode, bNode)
+	if edge := NewEdgeIfMissing(l.Graph, i, aNode, bNode, l.NewEdgeMetadata()); edge != nil {
+		return append(inputEdges, edge)
+	}
+	return inputEdges
+}
+
+// GetABLinks implementing graph.Linker
+func (l *ABLinker) GetABLinks(aNode *graph.Node) (edges []*graph.Edge) {
+	namespace, _ := aNode.GetFieldString(MetadataField("Namespace"))
+	if a := l.ACache.GetByNode(aNode); a != nil {
+		for _, b := range l.BCache.getByNamespace(namespace) {
+			uid := b.(metav1.Object).GetUID()
+			if bNode := l.Graph.GetNode(graph.Identifier(uid)); bNode != nil {
+				if l.AreLinked(a, b) {
+					edges = l.AppendNewEdge(edges, aNode, bNode)
+				}
+			}
+		}
+	}
+	return
+}
+
+// GetBALinks implementing graph.Linker
+func (l *ABLinker) GetBALinks(bNode *graph.Node) (edges []*graph.Edge) {
+	namespace, _ := bNode.GetFieldString(MetadataField("Namespace"))
+	if b := l.BCache.GetByNode(bNode); b != nil {
+		for _, a := range l.ACache.getByNamespace(namespace) {
+			uid := a.(metav1.Object).GetUID()
+			if aNode := l.Graph.GetNode(graph.Identifier(uid)); aNode != nil {
+				if l.AreLinked(a, b) {
+					edges = l.AppendNewEdge(edges, aNode, bNode)
+				}
+			}
+		}
+	}
+	return
+}
+
+// NewABLinker create and initialize an ABLinker based linker
+func NewABLinker(g *graph.Graph, aManager, aType, bManager, bType string, outerLinker graph.Linker, areLinked AreLinked) probe.Probe {
+	aProbe := GetSubprobe(aManager, aType)
+	bProbe := GetSubprobe(bManager, bType)
+
+	if aProbe == nil || bProbe == nil {
+		return nil
+	}
+
+	innerLinker := outerLinker.(ABLinkerInterface).GetABLinker()
```

Comment:
Why do you pass a graph.Linker to then cast it to ABLinkerInterface to then call GetABLinker to get the value you had in the first place ?

Reply:
I now longer pass an ABLinker - I create one within NewABLinker() func